src/codesearch.cc

```diff
@@ -352,6 +368,49 @@ class searcher {
     friend class code_searcher::search_thread;
 };
 
+class filename_searcher {
+public:
+    filename_searcher(const code_searcher *cc,
+                      const query &q) :
+        cc_(cc), query_(&q), limiter_(q.max_matches) {}
+
+    void operator()(indexed_file *file);
+
+protected:
+    const code_searcher *cc_;
+    const query *query_;
+    thread_queue<file_result*> queue_;
+    atomic_int matches_;
+    search_limiter limiter_;
+
+    friend class code_searcher::search_thread;
+};
+
+void filename_searcher::operator()(indexed_file *file)
+{
+    if (limiter_.why())
```

Comment:
Looking at the code here:
 * This particular use is no longer inside a loop after adding an index on filenames. I don't think I'm going to bother to expunge it from the history. (Unless you'd prefer to squash the commits so that there's one commit that changes the RPC interface and sets up the index all at once.)
 * Another potential source of slowness when using `exit_early()` is the `matches_.load()`. I don't know how slow an `atomic_int::load()` is, but it seems like it might be a bit expensive. So I'll go ahead and add another version of `exit_early()` that doesn't include the match-limit check, for use when we know the match limit hasn't changed.